---
title: "Pruebas OncoSimulR"
date: "27/12/2019"
output: html_document
---
<style>
body {
text-align: justify}
</style>
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r message = FALSE, warning = FALSE}
suppressPackageStartupMessages(library("OncoSimulR"))
suppressPackageStartupMessages(library("ggplot2"))

options(stringsAsFactors = FALSE) ## get rid of the messages


## Hay que mejorarlas
# Cambiar nombre etiquetas
simul_boxplot<- function(df) {
  e <- ggplot(stack(df), aes(x = ind, y = values))
  e + geom_boxplot(aes(fill = ind)) +
    stat_summary(fun.y = mean, geom = "point",
                 shape = 18, size = 2.5, color = "#FC4E07") 
}

compositionPop <- function(objPop) {
  condi <- c("WT", objPop[[1]]$geneNames)
  listPop <- lapply(objPop, function(x) tail(x[[1]], 1)[1, -1])
  dfPop <- data.frame(matrix(unlist(listPop), ncol = length(condi), byrow = TRUE))
  colnames(dfPop) <- condi
  simul_boxplot(dfPop)
  # dfPopSummary <- apply(dfPop, 2, mean)
  # return(data.frame("Mean" = dfPopSummary))
}

```

Función modificada para realizar la simulación varias veces y representar los resultados. Lo único que he modificado es el empleo de "sapply" en vez de "lapply" porque así ya se te devuelve directamente como matriz y es más cómodo que lapply. Luego, he intentado que dfPop sea un data frame con solo 2 columnas (una correspondiente al genotipo y otra al nº de individuos obtenidos al final de cada simulación) para que tenga pinta de data frame y pueda ser empleado por ggplot sin la necesidad del stack. Es simplemente otra versión para que el objeto generado esté organizado como un data frame (por si lo ve Ramón). Pero hace lo mismo. Simplemente.

```{r message = FALSE, warning = FALSE}
## Plot boxplot
simul_boxplot2 <- function(df) {
  e <- ggplot(df, aes(x = Genotype, y = N))
  e + geom_boxplot(aes(fill = Genotype)) +
    stat_summary(fun.y = mean, geom = "point",
                 shape = 18, size = 2.5, color = "#FC4E07") 
  }

## Extract and create a data frame with results from several simulations
compositionPop2 <- function(objPop) {
  clon_labels <- c("WT", objPop[[1]]$geneNames)
  listPop <- vapply(objPop, function(x) tail(x[[1]], 1)[1, -1], 
                    as.double(1:length(clon_labels)))
  dfPop <- data.frame("Genotype" = rep(clon_labels, 
                                       length(listPop)/length(clon_labels)),
                      "N" = c(listPop))
  simul_boxplot2(dfPop)
  }

```

## Leyenda
Estas líneas de código corresponden a generar la leyenda fuera del plot. No sé si se podría hacer automático con otra cosa, pero bueno, en cualquier caso lo único que teneis que hacer es ejecutar el trace() y copiar tal cual el código siguiente. Corresponde con la misma función, pero con las modificaciones que permiten mover la leyenda. De hecho, ahora probaré si puedo hacer que podamos moverla con argumentos, pero para eso hay que modificar otra función también.

```{r}

## Habría que borrar lo de arriba porque están incluidas estas funciones en el script
source("./Funciones_plot_markdown.R")
```

```{r}

# trace(OncoSimulR:::plotClonesSt, edit = T)
# 
# function (z, ndr, show = "drivers", na.subs = TRUE, log = "y", 
#     lwd = 1, lty = 1:8, col = 1:9, order.method = "as.is", stream.center = TRUE, 
#     stream.frac.rand = 0.01, stream.spar = 0.2, border = NULL, 
#     srange = c(0.4, 1), vrange = c(0.8, 1), type = "stacked", 
#     breakSortColors = "oe", colauto = TRUE, legend.ncols = "auto", 
#     lwdStackedStream = 1, xlab = "Time units", ylab = "Number of cells", 
#     ylim = NULL, xlim = NULL, ...) 
# {
#     y <- z$pops.by.time[, 2:ncol(z$pops.by.time), drop = FALSE]
#     if (type %in% c("stacked", "stream")) 
#         na.subs <- FALSE
#     if (na.subs) {
#         y[y == 0] <- NA
#     }
#     oo <- order(ndr)
#     y <- y[, oo, drop = FALSE]
#     ndr <- ndr[oo]
#     if (show == "drivers") {
#         col <- rep(col, length.out = (1 + max(ndr)))[ndr + 1]
#         lty <- rep(lty, length.out = ncol(y))
#     }
#     else {
#         if (length(col) < max(ndr)) 
#             warning("Repeating colors; you might want to", "pass a col vector of more elements")
#         col <- rep(col, length.out = (max(ndr)))[ndr]
#     }
#     if (type == "line") {
#         par(mar = c(4.1, 4.1, 4.2, 2.1))
#         matplot(x = z$pops.by.time[, 1], y = y, log = log, type = "l", 
#             col = col, lty = lty, lwd = lwd, xlab = xlab, ylab = ylab, 
#             ylim = ylim, xlim = xlim, ...)
#         box()
#         if (show == "genotypes") {
#             if (!inherits(z, "oncosimul2")) {
#                 ldrv <- genotypeLabel(z)
#             }
#             else {
#                 ldrv <- z$GenotypesLabels
#             }
#             ldrv[ldrv == ""] <- "WT"
#             ldrv[ldrv == " _ "] <- "WT"
#             if (legend.ncols == "auto") {
#                 if (length(ldrv) > 6) 
#                   legend.ncols <- 2
#                 else legend.ncols <- 1
#             }
#             par(xpd = TRUE)
#             legend(x = "top", title = "Genotypes", lty = lty, 
#                 inset = -0.29, col = col, lwd = lwd, legend = ldrv, 
#                 ncol = legend.ncols)
#         }
#     }
#     else {
#         ymax <- colSums(y)
#         if ((show == "drivers") || ((show == "genotypes") && 
#             (colauto))) {
#             cll <- myhsvcols(ndr, ymax, srange = srange, vrange = vrange, 
#                 breakSortColors = breakSortColors)
#         }
#         else {
#             cll <- list(colors = col)
#         }
#         x <- z$pops.by.time[, 1]
#         if (grepl("y", log)) {
#             stop("It makes little sense to do a stacked/stream", 
#                 "plot after taking the log of the y data.")
#         }
#         if (grepl("x", log)) {
#             x <- log10(x + 1)
#         }
#         if (type == "stacked") {
#             plot.stacked2(x = x, y = y, order.method = order.method, 
#                 border = border, lwd = lwdStackedStream, col = cll$colors, 
#                 log = log, xlab = xlab, ylab = ylab, ylim = ylim, 
#                 xlim = xlim, ...)
#         }
#         else if (type == "stream") {
#             plot.stream2(x = x, y = y, order.method = order.method, 
#                 border = border, lwd = lwdStackedStream, col = cll$colors, 
#                 frac.rand = stream.frac.rand, spar = stream.spar, 
#                 center = stream.center, log = log, xlab = xlab, 
#                 ylab = ylab, ylim = ylim, xlim = xlim, ...)
#         }
#         if (show == "drivers") {
#             if (legend.ncols == "auto") {
#                 if (length(cll$colorsLegend$Drivers) > 6) 
#                   legend.ncols <- 2
#                 else legend.ncols <- 1
#             }
#             legend(x = "topleft", title = "Number of drivers", 
#                 pch = 15, col = cll$colorsLegend$Color, legend = cll$colorsLegend$Drivers, 
#                 ncol = legend.ncols)
#         }
#         else if (show == "genotypes") {
#             if (!inherits(z, "oncosimul2")) {
#                 ldrv <- genotypeLabel(z)
#             }
#             else {
#                 ldrv <- z$GenotypesLabels
#             }
#             ldrv[ldrv == ""] <- "WT"
#             ldrv[ldrv == " _ "] <- "WT"
#             if (legend.ncols == "auto") {
#                 if (length(ldrv) > 6) 
#                   legend.ncols <- 2
#                 else legend.ncols <- 1
#             }
#             legend(x = "topleft", title = "Genotypes", pch = 15, 
#                 col = cll$colors, legend = ldrv, ncol = legend.ncols)
#         }
#     }
# }

```
  
## Primer ejemplo (Miguel): Local dispersal promotres biodiversity in a real fame of rock-paper-scissors.
Tiene que ver con E.coli y resistencia-sensibilidad a colicina. 
* (R): bacterias resistentes
* (C): bacterias productoras de colicina
* (S): bacterias sensibles

Situaciones.
1) En algunos casos, el ratio de crecimiento de R excederá al de C, pero será menor que el de S. Esto ocurre porque R evitan los costes competitivos de tener el plásmido col, pero sufren porque el receptor de colicina está involucrado en funciones cruciales para la bacteria. 
2) En otros casos, S puede desplazar a R (porque S tiene una ventaja en el ratio de crecimiento) y C puede desplazar a S (porque C mata a S).

```{r}
crs <- function (a, b, c){
  data.frame(Genotype = c("WT", "C", "R", "S"),
                    Fitness = c("1 - 10 * (f_1 + f_2 + f_3)",
                                paste0("1 + ", a, " * f_3 - ", b, " * f_2"),
                                paste0("1 + ", b, " * f_1 - ", c, " * f_3"),
                                paste0("1 + ", c, " * f_2 - ", a, " * f_1")
                    ))#Ecuaciones de interaccion entre los tipos
  }

afcrs <- allFitnessEffects(genotFitness = crs(2, 2, 2), #Par?metros de la interaccion
                          frequencyDependentFitness = TRUE,
                          frequencyType = "rel")

resultscrs <- oncoSimulIndiv(afcrs,
                              model = "McFL", #Se puede poner Exp, McFL, McFLD...
                              onlyCancer = FALSE,
                              finalTime = 100, #Tiempo de la simulacion
                              mu = 1e-4,
                              initSize = 40000, #Poblacion inicial
                              keepPhylog = TRUE,
                              seed = NULL,
                              errorHitMaxTries = FALSE,
                              errorHitWallTime = FALSE)

plot(resultscrs, show = "genotypes", type = "line")
```

## Segundo ejemplo (Álvaro): The Lotka-Volterra model of competition between two competing species

Species do not exist in isolation of one another. The simple models of exponential and logistic growth fail to capture the fact that the growth of a species must depend on its interactions with other species (Otto and Day 2007). Including all possible factors in a model is very complex. However, there are models that focus on specific interactions and how they affect population growth, for example, Lotka-Volterra model.

The Lotka-Volterra model extends the logistic model by allowing multiple species to compete for resources, resulting in both inter-specific and intra-specific competition. In this part we will consider only two competing species (S1 and S2). 

Competition model describe how the number of individuals of each species chenges when other species uses the same resource. 

The parameters considered in Lotka-Volterra model are:
  $n$ -> number of individuals
  $r$ ->  Intrinsic growth rates
  $K$ -> Carrying capacities (max capacity of the environment)
  $a_{ij}$ -> competition coefficient. The effect of species j on species i

With these parameters, the continuous-time differential equations for the Lotka-Volterra model are:

$$ \frac{dn_1}{dt}= r_1n_1(t)\left(1-\frac{n_1(t)+a_{12}n_2(t)}{K_1}\right)$$
$$ \frac{dn_2}{dt}= r_2n_2(t)\left(1-\frac{n_2(t)+a_{21}n_1(t)}{K_2}\right)$$

As it is said above, the signs of $a_{12}$ and $a_{21}$ therefore reflect the relationship between the two species:
      a_12   |   a_21   | Relationship
    -------- | -------- | -------------
       -     |    -     | Mutualistic
       -     |    0     | Commensal
       0     |    -     | Commensal
       +     |    -     | Parasitic
       -     |    +     | Parasitic
       +     |    +     | Competitive
       +     |    0     | Amensalism
       0     |    +     | Amensalism

Next, to show the effect of competition coefficients and competition models we will model the scenarios proposed in the book "Analysis and Management of Animal Populations" (Williams, Nichols and Conroy, 2010, 168-171). For the following examples S1 and S2 will be the two species used, and can be interpreted as cellular clones competing for cellular resources or multicellular organisms competing for environmental resources (sexual or asexual reproduction is not relevant as it is incorporated in the birth rate). 

S2 in these examples has a higher rate of growth than S1 ($r2 > r1$) and higher carrying capacity ($K2 > K1$). However, S1 has a stronger competitive effect on S2 ($a_{21} > a_{12}$).

Before seeing the examples, we have to keep in mind that in these models we treat resources as constant and the competition comes from the fight for resources, in any case is one of the species the resource. 

```{r}
## Lotka-Volterra expressions for competition models
S1S2_LV <- function(r1, r2, K1, K2, a_12, a_21, awt = 1,
                    gt = c("WT", "S1", "S2")) {
  data.frame(Genotype = gt,
             Fitness = c(
               paste0("max(0.1, 1 - ", awt, " * (n_2 + n_1))"),
               paste0("1 + ", r1,
                      " * ( 1 - (n_1 + ", a_12, " * n_2)/", K1,
                      ")"),
               paste0("1 + ", r2,
                      " * ( 1 - (n_2 + ", a_21, " * n_1)/", K2,
                      ")")
             ))
}

## Show expressions for birth rates
S1S2_LV("r1", "r2", "K1", "K2", "a_12", "a_21", "awt")

```

### Case 1: No competition

There is no competition, the competition coefficients = 0. We see that the populations manage to reach their $K$ (maximum supported population). Due to S2 has greater capacity and growth rate, it reaches a population size greater than S1.
```{r, message=FALSE}
## Case 1: No competition effects
## K2 > K1; r2 > r1 (species 2 has greater capacity and growth rate)
S1S2_competition <-allFitnessEffects(
    genotFitness = S1S2_LV(1.2, 1.7, 350, 400, 0, 0, ## a_12, a_21 are 0
              gt = c("WT", "S1", "S2")),
    frequencyDependentFitness = TRUE,
    frequencyType = "abs", ## frequency expressed in absolute terms
    spPopSizes = c(50, 50, 50)) ## Same population, lower than K

## Plot fitness landscape of genotype "S1, S2" evaluation
data.frame("S1_fitness" = evalGenotype(genotype = "S1", 
                                       fitnessEffects = S1S2_competition), 
           "S2_fitness" = evalGenotype(genotype = "S2", 
                                       fitnessEffects = S1S2_competition))

S1S2_simulation <- oncoSimulPop(5, S1S2_competition, ## 10 simulations
                              model = "Exp",
                              onlyCancer = FALSE,
                              finalTime = 500,
                              mu = 1e-1, ## Allows appearance of S1 and S2
                              initSize = 4000,
                              seed = NULL,
                              errorHitMaxTries = FALSE,
                              errorHitWallTime = FALSE)



## Fitness values as an example
plot(S1S2_simulation[[1]], show = "genotypes", type = "line", 
     xlim = c(40, 500), 
     ylab = "Number of individuals",
     main = "No competition effect", 
     col=c("black","salmon", "darkgreen"),
     font.main=2, font.lab=2,
     cex.main=2, cex.lab=1.2,
     las = 1)
```

It is also important to show a boxplot with the results of different simulations to check the reproducibility of the model. 

```{r, message=FALSE}
## Agreement: Show boxplot from several simulations. 
compositionPop2(S1S2_simulation)
```


Before continuing with different scenarios, it is interesting to observe how the fitness landscape of the species S1 and S2 is influenced by the load capacity (K). In the absence of competition (exponential model), a fitness value larger than 1 indicates growth; a fitness value smaller than 1 indicates decline. Thus, when fitness is studied for populations of S1 and S2 with the size equal to K1 and K2 respectively, it is observed that the fitness is 1.
```{r, message=F}
## Case 1: No competition effects
S1S2_competition <-allFitnessEffects(
    genotFitness = S1S2_LV(1.2, 1.7, 350, 400, 0, 0,
              gt = c("WT", "S1", "S2")),
    frequencyDependentFitness = TRUE,
    frequencyType = "abs", 
    spPopSizes = c(5, 350, 400)) ## spPopSizes equal to K1 and K2 respectively

## Fitness values of genotype "S1, S2" evaluation
data.frame("S1_fitness" = evalGenotype(genotype = "S1", 
                                       fitnessEffects = S1S2_competition), 
           "S2_fitness" = evalGenotype(genotype = "S2", 
                                       fitnessEffects = S1S2_competition))
```

In contrast, when spPopsizes is greater than K1 or K2 a population decrease is observed because the environment cannot support that size. 
```{r, message=F}
## Case 1: No competition effects
S1S2_competition <-allFitnessEffects(
    genotFitness = S1S2_LV(1.2, 1.7, 350, 400, 0, 0,
              gt = c("WT", "S1", "S2")),
    frequencyDependentFitness = TRUE,
    frequencyType = "abs",
    spPopSizes = c(5, 500, 500)) ## spPopSizes higher than K1 and K2 respectively

## Fitness valuesof genotype "S1, S2" evaluation
data.frame("S1_fitness" = evalGenotype(genotype = "S1", 
                                       fitnessEffects = S1S2_competition), 
           "S2_fitness" = evalGenotype(genotype = "S2", 
                                       fitnessEffects = S1S2_competition))
```

### Case 2: Stable coexistence
As described in page 166 from the book "Analysis and Management of Animal Populations" (Williams, Nichols and Conroy, 2010), if the numerators of the equilibrium formulas are positive, then the populations can coexist in equilibrium:

  $K1 > a_{12}*K2$   
  $K2 > a_{21}*K1$

In this case, steady-state population (ESS) levels are simply the carrying capacieties K1 and K2, reduced by amounts $a_{12}*K2$ and $a_{21}*K1$, repectively, and scales by $1-a_{12}a_{21}$. Both populations converge to the equilibrium population levels irrespective of the initial population sices. 

Considering: 
  $K1$ -> 350
  
  $K2$ -> 400
  
  $a_{12}$ -> 0.25
  
  $a_{21}$ -> 0.5
  
$350 > 0.25*400$ and $400 > 0.5*350$

In this example we will see the stable coexistance of S1 and S2, which may change the final size of each population  from the previous example. 

```{r, message=FALSE}
## Case 2: Competition with Stable Coexistance
## K2 > K1; r2 > r1 (specie 2 has greater capacity and growth rate)
## a21 > a12 (specie 1 has more effect on specie 2)
S1S2_competition <-allFitnessEffects(
    genotFitness = S1S2_LV(1.2, 1.7, 350, 400, 0.25, 0.5,
              gt = c("WT", "S1", "S2")),
    frequencyDependentFitness = TRUE,
    frequencyType = "abs")


S1S2_simulation <- oncoSimulPop(5, 
                                S1S2_competition,
                              model = "Exp",
                              onlyCancer = FALSE,
                              finalTime = 500,
                              mu = 1e-1,
                              initSize = 200,
                              keepPhylog = TRUE,
                              seed = NULL,
                              errorHitMaxTries = FALSE,
                              errorHitWallTime = FALSE)

## Plot first trajectory as an example
plot(S1S2_simulation[[1]], show = "genotypes", type = "line", 
     xlim = c(40, 500), 
     ylab = "Number of individuals", 
     main = "Competition with Stable Coexistance", 
     col=c("black","salmon", "darkgreen"),
     font.main=2, font.lab=2,
     cex.main=2, cex.lab=1.2,
     las = 1)

## Agreement: Show boxplot from several simulations. 
compositionPop2(S1S2_simulation)

```

As the previous case, if we analyze the fitness values for S1 and S2 we could check that initial population sizes lower than $n_1(t)$ and $n_2(t)$ lead to increases toward the steady-states values, this means fitness values higher than 1.
```{r, message=FALSE}
S1S2_competition <-allFitnessEffects(
    genotFitness = S1S2_LV(1.2, 1.7, 350, 400, 0.25, 0.5,
              gt = c("WT", "S1", "S2")),
    frequencyDependentFitness = TRUE,
    frequencyType = "abs",
    spPopSizes = c(5, 50, 50)) ## lower initial sizes than ESS

## Fitness values of genotype "S1, S2" evaluation
data.frame("S1_fitness" = evalGenotype(genotype = "S1",
                                       fitnessEffects = S1S2_competition), 
           "S2_fitness" = evalGenotype(genotype = "S2", 
                                       fitnessEffects = S1S2_competition))
```

Furthermore, initial population sizes higher than $n_1(t)$ and $n_2(t)$ lead to decreases toward the steady-states values, this means fitness values lower than 1.

```{r, message=FALSE}
S1S2_competition <-allFitnessEffects(
    genotFitness = S1S2_LV(1.2, 1.7, 350, 400, 0.25, 0.5,
              gt = c("WT", "S1", "S2")),
    frequencyDependentFitness = TRUE,
    frequencyType = "abs",
    spPopSizes = c(5, 410, 410)) ## higher initial sizes than ESS

## Fitness values of genotype "S1, S2" evaluation
data.frame("S1_fitness" = evalGenotype(genotype = "S1",
                                       fitnessEffects = S1S2_competition), 
           "S2_fitness" = evalGenotype(genotype = "S2", 
                                       fitnessEffects = S1S2_competition))
```

## Case 3: Competitive Exclusion
If only one of the two conditions $K1 > a_{12}*K2$ or $K2 > a_{21}*K1$ is achivedm the corresponding population eventually approaches its caryying capacity (irrespective of population initial size) and the other population is driven to extinction.

Considering: 
  $K1$ -> 350
  $K2$ -> 400
  $a_{12}$ -> 0.5
  $a_{21}$ -> 1.2
  
$350 > 0.5*400$ and $400 > 1.2*350$

```{r, message=FALSE}
## Test 3: Estable competition but more aggressive
# Vemos que las dos especies no son compatibles y el estado no es estable
## Ambas impactan bastante sobre la otra, pero pueden coexitir
## Se alejan del K original
S1S2_competition <-
  allFitnessEffects(
    genotFitness =
      S1S2_LV(1, 1.7, 350, 400, 0.5 , 1.2,
              gt = c("WT", "S1", "S2")),
    frequencyDependentFitness = TRUE,
    frequencyType = "abs")

S1S2_simulation <- oncoSimulPop(5, S1S2_competition,
                              model = "Exp",
                              onlyCancer = FALSE,
                              finalTime = 200,
                              mu = 1e-1,
                              initSize = 200,
                              keepPhylog = TRUE,
                              seed = NULL,
                              errorHitMaxTries = FALSE,
                              errorHitWallTime = FALSE)

## Plot first trajectory as an example
plot(S1S2_simulation[[1]], show = "genotypes", type = "line", 
     xlim = c(10, 200), 
     ylab = "Number of individuals", 
     main = "Competitive Exclusion", 
     col=c("black","salmon", "darkgreen"),
     font.main=2, font.lab=2,
     cex.main=2, cex.lab=1.2,
     las = 1)

## Agreement: Show boxplot from several simulations. 
compositionPop2(S1S2_simulation)
```


### Case 4: Unstable Population Equilibrium
If both condition are:
  $K1 < a_{12}*K2$   
  $K2 < a_{21}*K1$

The population sizes never reach to equilibrium, so always one of the competitors will be extinct. The extincted competitor will depend on:
$$n_2 > \frac{K_2-a_{12}K_1}{K_1-a_{21}K_2}\ n_1$$
where $n_1$ will be extincted. Or:

$$n_2 < \frac{K_2-a_{12}K_1}{K_1-a_{21}K_2}\ n_1 $$
where $n_2$ will be extincted.

This is because, as we explained before, the steady-state population levels scaled by $1-a_{12}a_{21}$. In these scenarios: $a_{12}a_{21}>1$, so the steady-state population will never be reached.

In the first example we have:
  $K1$ -> 350
  $K2$ -> 400
  $a_{12}$ -> 1
  $a_{21}$ -> 1.4

$n_2 > 1.8n_1$

Since the mutation rate for S1 and S2 is the same (mu = 0.1), we can conclude that the initial population of n_1 and _n2 will be approximately equal to the same t. Thus the right-hand term is larger than the left-hand term. Therefore, S2 will be extincted:

```{r, message=F}
## Test 4: Unestable competition
# Vemos que las dos especies no son compatibles y el estado no es estable
## Una de ellas se extingue, y la otra llega a un valor cercano a la K.
## Sobrevive S1 que es la que m?s impacto tiene sobre la otra (a_21 > a_12)

# Aquí gana N2
S1S2_competition <-
  allFitnessEffects(
    genotFitness =
      S1S2_LV(1.2, 1.7, 350, 400, 1, 1.4, ## notice coefficients
              gt = c("WT", "S1", "S2")),
    frequencyDependentFitness = TRUE,
    frequencyType = "abs")

S1S2_simulation <- oncoSimulPop(2, S1S2_competition,
                              model = "Exp",
                              onlyCancer = FALSE,
                              finalTime = 500,
                              mu = 1e-1,
                              initSize = 4000,
                              keepPhylog = TRUE,
                              seed = NULL,
                              errorHitMaxTries = FALSE,
                              errorHitWallTime = FALSE)

## Plot first trajectory as an example
plot(S1S2_simulation[[1]], show = "genotypes", type = "line", 
     xlim = c(15, 500), 
     ylab = "Number of individuals", 
     main = "Unestable competition", 
     col=c("black","salmon", "darkgreen"),
     font.main=2, font.lab=2,
     cex.main=2, cex.lab=1.2,
     las = 1)

## Agreement: Show boxplot from several simulations. 
compositionPop2(S1S2_simulation)
```

In contrast, in the next example we have:
  $K1$ -> 350
  $K2$ -> 400
  $a_{12}$ -> 1
  $a_{21}$ -> 1.2
  

$n_2 < 0.4n_1$
Thus the right-hand term is lower than the left-hand term. Therefore, S1 will be extincted:

```{r, message=F}
## Test 4: Unestable competition
S1S2_competition <-
  allFitnessEffects(
    genotFitness =
      S1S2_LV(1.2, 1.7, 350, 400, 1, 1.2, ## notice coefficients
              gt = c("WT", "S1", "S2")),
    frequencyDependentFitness = TRUE,
    frequencyType = "abs")

S1S2_simulation <- oncoSimulPop(2, S1S2_competition,
                              model = "Exp",
                              onlyCancer = FALSE,
                              finalTime = 500,
                              mu = 1e-1,
                              initSize = 4000,
                              keepPhylog = TRUE,
                              seed = NULL,
                              errorHitMaxTries = FALSE,
                              errorHitWallTime = FALSE)

## Plot first trajectory as an example
plot(S1S2_simulation[[1]], show = "genotypes", type = "line", 
     xlim = c(10, 500), 
     ylab = "Number of individuals", 
     main = "Unestable competition", 
     col=c("black","salmon", "darkgreen"),
     font.main=2, font.lab=2,
     cex.main=2, cex.lab=1.2,
     las = 1)

## Agreement: Show boxplot from several simulations. 
compositionPop2(S1S2_simulation)
```

## Tercer ejemplo (Diego): Investigating prostate cancer tumour–stroma interactions: clinical and biological insights from an evolutionary game (<http://dx.doi.org/10.1038/bjc.2011.517>)

**Nota:** Este ejemplo simula la relación entre dos tipos de poblaciones tumorales y una población no tumoral. No simula relaciones de epistasia entre genes, etc.

### Introducción (del paper)
In this paper we introduce an Evolutionary Game Theory (EGT) model that studies the evolution of three different cell populations over time: stromal cells, a dependant tumour phenotype capable of co-opting stromal cells to support its growth, and an independent tumour phenotype that does not require microenvironmental support, be it stromal associated or not.
The payoff is fitness in the Darwinian sense: more average reproductive success. In this context, interactions between the players are important, be it in a cooperative or competitive manner, as they determine whether it will become a larger share of the population (Sigmund and Nowak, 1999). Fitness is also population dependant, so a given subpopulation might become more (or less) fit if the numbers of a different subpopulation with whom it interacts go up (or down).


**Más papers con posibles modelos (ya hechos)**
EGT has been used to explore various aspects of cancer (Tomlinson and Bodmer, 1997), including glioma progression (Basanta et al, 2008, 2011), the influence of the tumour – host interface in colorectal carcinogenesis (Gatenby and Vincent, 2003), the role of phenotypic variability in multiple myeloma (Dingli et al, 2009), and the evolution of a number of phenotypic traits such as motility and invasion (Mansury et al, 2006; Bellomo et al, 2008) or microenvironmental independence (Anderson et al, 2009).


### Modelo
There are only four variables in the model,which is the minimum necessity to consider how the costs and benefits of either relaying on the stroma and the stromal cells (a, b, r == p)) or on the other hand being independent (g) affect the outcome of prostate cancer.

**Attention:** The payoff values are normalised in the range (0:1) so 1 represents the maximum fitness for any given phenotype.

Tenemos, en primer lugar, tres tipos celulares:

1. S: células estromales.  
2. D: células tumorales dependientes del microambiente.  
3. I: células tumorales independientes del microambiente.  

Las relaciones entre sí en términos de costes y beneficios son:

* $\alpha$ (a): beneficio de la cooperación entre S y D y entre D y D.  
* $\beta$ (b): coste de extraer recursos del microambiente.  
* $\gamma$ (g): coste de ser independiente del microambiente (no lo usan en las fórmulas, repasar porque probablemente lo usen en algún punto del paper).  
* $\rho$ (p): beneficio de D por factores de crecimiento paracrinos producidos por I.  

### Explicación
Las células I son independientes, asumiéndose que tienen un fitness constante (1 - $\gamma$).  
Las células D son dependientes del microambiente para obtener sus recursos, por lo que los costes asociados a la extracción de éstos se representan con $\beta$. Microambientes con pocos recursos estarán representados, por tanto, por altos valores de $\beta$ (y viceversa). Como las células I producen espacio y factores de crecimiento compartibles, el modelo asume que las células D obtienen una ventaja adaptativa de sus interacciones con las células I, lo cual está representado por la variable $\rho$ (p).  
Por otra parte, células D que interactúen con otras células D tendrán problemas compartiendo los recursos existentes en el microambiente y por lo tanto, se asume que tienen el doble de costo $\beta$ por depender del microambiente para su supervivencia, teniendo un fitness de 1 - 2 * $\beta$.  
También se consideran poblaciones de células estromales (S) que pueden interactuar con el tumor. Las células S conservan la capacidad de sufrir una rápida proliferación, pero normalmente tienen un crecimiento quiescente con bajos ratios de proliferación y muerte, por lo que se suponen con un fitness igual a cero. Sin embargo, bajo ciertas circunstancias, las células estromales son susceptibles de cooperar con ciertos fenotipos tumorales. En esta situacuón, células tumorales y estromales producen factores que pueden estimular el crecimiento de las poblaciones entre sí de forma mutualista. Esta relación mutualista está representada mediante $\alpha$. Bajos valores de $\alpha$ representan tumores donde el estroma no puede cooperar.  

```{r example3_equations}

# Con WT y S
# El problema aquí es que estoy modelando que S proceden de una población inicial WT, lo que no es cierto. Creo que sería mejor poner S como WT.
pan_rel_01 <- function (a, b, p, g, gt = c("WT", "S", "I", "D")) {
    data.frame(Genotype = gt,
               Fitness = c("1",
                      paste0("1 + ", a, " * f_3"),
                      paste0("1 + 1 - ", g),
                      paste0("1 + ", "(1 - f_2 - f_3) * (1 - ", b, " + ", a,
                          ") + f_2 * (1 - ", b, " + ", p, ") + f_3 * (1 - 2 * ",
                          b, ") + 1 - ", b, " + ", a, " + f_2 * (", p, " - ",
                          a, ") - f_3 * (", b, " + ", a, ")")
                    )
             )
}

# Con S como WT
pan_rel_02 <- function (a, b, p, g, gt = c("WT", "I", "D")) {
    data.frame(Genotype = gt,
               Fitness = c(paste0("1 + ", a, " * f_2"),
                      paste0("1 + 1 - ", g),
                      paste0("1 + ", "(1 - f_1 - f_2) * (1 - ", b, " + ", a,
                          ") + f_1 * (1 - ", b, " + ", p, ") + f_2 * (1 - 2 * ",
                          b, ") + 1 - ", b, " + ", a, " + f_1 * (", p, " - ",
                          a, ") - f_2 * (", b, " + ", a, ")")
                    )
             )
}

pan_rel_01("a", "b", "p", "g")

```

### Pruebas en diferentes escenarios.
One may apply the replicator equations described in the previous section to study the temporal evolution of the different populations in a number of scenarios. These scenarios are characterised by the four variables of the model: a, the mutual benefit between D cells and co-opted stroma; r, the benefit that D cells derive from coexisting with I cells (which can produce growth factors and space); b, the fitness cost of relying on a microenvironment that might not be able to produce resources that I cells can produce on its own and g, the cost that I cells have to incur to become independent from the microenvironment.

* $\alpha$: beneficio mutuo entre células D y cooperación con células S.
* $\rho\varrho$: beneficio de células D por coexistir con células I (que producen factores de crecimiento y espacio).
* $\beta$: coste en el fitness de depender de un microambiente que no es capaz de producir recursos que las células Ip ueden producir por sí mismas.
* $\gamma$: costo en el que deben incurrir las células I para ser independientes del microambiente. 



#### **Primer escenario.**

* a = 0.5: cooperación entre D-D y D-S intermedia. 
* b = 0.7: coste de la obtención de recursos: microambiente relativamente pobre. 
* p = 0.1: beneficio bajo de células D por coexistir con I. 
* g = 0.8: coste de ser independiente alto. 

Un valor suficientemente alto de a y suficientemente bajo de p se traduce en que las células D obtienen un beneficio mayor de su interacción con las células S que de su relación con las células I. Dado al alto coste de independencia g, no es sorprendente que las células I acaben extinguiéndose y el tumor acabe estando formado por los otros dos tipos celulares. 

##### Versión con S separado de WT
```{r example3_scen1_1}

scen1 <- allFitnessEffects(genotFitness = pan_rel_01(a = 0.5, b = 0.7,
                                                     p = 0.1, g = 0.8),
                          frequencyDependentFitness = TRUE,
                          frequencyType = "rel")

simulScen1 <- oncoSimulIndiv(scen1,
                              model = "McFL",
                              onlyCancer = FALSE,
                              finalTime = 100,
                              mu = 1e-4,
                              initSize = 40000,
                              keepPhylog = TRUE,
                              seed = NULL,
                              errorHitMaxTries = FALSE,
                              errorHitWallTime = FALSE)

plot(simulScen1, show = "genotypes", type = "line")
```

##### Versión con WT como S
Es solo un ejemplo, el resultado es el mismo respecto al primero, pero no sé qué es mejor.

```{r example3_scen1_2}

scen1 <- allFitnessEffects(genotFitness = pan_rel_02(a = 0.5, b = 0.7,
                                                     p = 0.1, g = 0.8),
                          frequencyDependentFitness = TRUE,
                          frequencyType = "rel")

simulScen1 <- oncoSimulIndiv(scen1,
                              model = "McFL",
                              onlyCancer = FALSE,
                              finalTime = 100,
                              mu = 1e-4,
                              initSize = 4000,
                              keepPhylog = TRUE,
                              seed = NULL,
                              errorHitMaxTries = FALSE,
                              errorHitWallTime = FALSE)

plot(simulScen1, show = "genotypes", type = "line")
```

#### **Segundo escenario.**
* a = 0.3: cooperación entre D-D y D-S baja 
* b = 0.7: microambiente relativamente pobre.
* p = 0.1: beneficio bajo de células D por coexistir con I.
* g = 0.7: coste menor al escenario 1 de ser independiente.

Al facilitar la posibilidad de que las células I puedan ser independientes (menor coste en su fitness), en lugar de extinguirse como en el primer escenario acaban prevaleciendo junto con las células D (a pesar del bajo beneficio su cooperación con I).

**Nota**
Al hacer varias iteraciones, las cosas cambian. Igual habría que utilizar oncoSimulSample para hacer un promedio de cada escenario. 
Fijarse en los finalTime de cada versión: como el número de poblaciones es diferente, las relaciones también, teniéndolo que cambiar. 

##### Versión con S separado de WT

```{r}
scen1 <- allFitnessEffects(genotFitness = pan_rel_01(a = 0.3, b = 0.7,
                                                     p = 0.1, g = 0.7),
                          frequencyDependentFitness = TRUE,
                          frequencyType = "rel")

simulScen1 <- oncoSimulIndiv(scen1,
                              model = "McFL",
                              onlyCancer = FALSE,
                              finalTime = 150,
                              mu = 1e-4,
                              initSize = 4000,
                              keepPhylog = TRUE,
                              seed = NULL,
                              errorHitMaxTries = FALSE,
                              errorHitWallTime = FALSE)

plot(simulScen1, show = "genotypes", type = "line")
```

##### Versión con WT como S

```{r}
scen1 <- allFitnessEffects(genotFitness = pan_rel_02(a = 0.3, b = 0.7,
                                                     p = 0.1, g = 0.7),
                          frequencyDependentFitness = TRUE,
                          frequencyType = "rel")


simulScen1 <- oncoSimulIndiv(scen1,
                              model = "McFL",
                              onlyCancer = FALSE,
                              finalTime = 200,
                              mu = 1e-4,
                              initSize = 4000,
                              keepPhylog = TRUE,
                              seed = NULL,
                              errorHitMaxTries = FALSE,
                              errorHitWallTime = FALSE)

plot(simulScen1, show = "genotypes", type = "line")
```

#### **Tercer escenario.**
* a = 0.2: cooperación entre D-D y D-S baja. 
* b = 0.35: microambiente relativamente rico, beneficioso para células D.
* p = 0.1: beneficio bajo de células D por coexistir con I.
* g = 0.3: coste de ser independiente muy bajo.

Situación extrema: a pesar de que el coste de la independencia es bastante bajo (solo 0.3 en relación a escenarios anteriores), el hecho de que el ambiente sea moderadamente rico (b = 0.35), hace que las células D (dependientes del microambiente) prevalezcan sobre el resto, incluso sobre las I. Con hacer solo un poco más pobre el microambiente (b = 0.40), las células D y las I sobreviven, cooperando entre sí (Cuarto escenario).

##### Versión con S separado de WT

```{r}
scen1 <- allFitnessEffects(genotFitness = pan_rel_01(a = 0.2, b = 0.35,
                                                     p = 0.1, g = 0.3),
                          frequencyDependentFitness = TRUE,
                          frequencyType = "rel")

simulScen1 <- oncoSimulIndiv(scen1,
                              model = "McFL",
                              onlyCancer = FALSE,
                              finalTime = 100,
                              mu = 1e-4,
                              initSize = 4000,
                              keepPhylog = TRUE,
                              seed = NULL,
                              errorHitMaxTries = FALSE,
                              errorHitWallTime = FALSE)

plot(simulScen1, show = "genotypes", type = "line")
```

##### Versión con WT como S

```{r}
scen1 <- allFitnessEffects(genotFitness = pan_rel_02(a = 0.2, b = 0.35,
                                                     p = 0.1, g = 0.3),
                          frequencyDependentFitness = TRUE,
                          frequencyType = "rel")

simulScen1 <- oncoSimulIndiv(scen1,
                              model = "McFL",
                              onlyCancer = FALSE,
                              finalTime = 100,
                              mu = 1e-4,
                              initSize = 4000,
                              keepPhylog = TRUE,
                              seed = NULL,
                              errorHitMaxTries = FALSE,
                              errorHitWallTime = FALSE)

plot(simulScen1, show = "genotypes", type = "line")
```

#### **Cuarto escenario.**

##### Versión con S separado de WT

```{r}
scen1 <- allFitnessEffects(genotFitness = pan_rel_01(a = 0.2, b = 0.4,
                                                     p = 0.1, g = 0.3),
                          frequencyDependentFitness = TRUE,
                          frequencyType = "rel")


simulScen1 <- oncoSimulIndiv(scen1,
                              model = "McFL",
                              onlyCancer = FALSE,
                              finalTime = 100,
                              mu = 1e-4,
                              initSize = 4000,
                              keepPhylog = TRUE,
                              seed = NULL,
                              errorHitMaxTries = FALSE,
                              errorHitWallTime = FALSE)
simulScen1
plot(simulScen1, show = "genotypes", type = "line")
```

##### Versión con WT como S

```{r}
scen1 <- allFitnessEffects(genotFitness = pan_rel_02(a = 0.2, b = 0.4,
                                                     p = 0.1, g = 0.3),
                          frequencyDependentFitness = TRUE,
                          frequencyType = "rel")

simulScen1 <- oncoSimulIndiv(scen1,
                              model = "McFL",
                              onlyCancer = FALSE,
                              finalTime = 15,
                              mu = 1e-4,
                              initSize = 4000,
                              keepPhylog = TRUE,
                              seed = NULL,
                              errorHitMaxTries = FALSE,
                              errorHitWallTime = FALSE)

plot(simulScen1, show = "genotypes", type = "line")
```

### Simulación con x réplicas del escenario 1

```{r}
scen1 <- allFitnessEffects(genotFitness = pan_rel_01(a = 0.5, b = 0.7,
                                                     p = 0.1, g = 0.8),
                          frequencyDependentFitness = TRUE,
                          frequencyType = "rel")

# Meter paralelización
simulScen1 <- oncoSimulPop(20,
                              scen1,
                              model = "McFL",
                              onlyCancer = FALSE,
                              mc.cores = 8, # paralelización
                              finalTime = 100,
                              mu = 1e-4,
                              initSize = 4000,
                              keepPhylog = TRUE,
                              seed = NULL,
                              errorHitMaxTries = FALSE,
                              errorHitWallTime = FALSE)

compositionPop(simulScen1)

```


### Simulación x repeticiones ejemplo Álex.

Me ha estado preguntado Álex respecto a su ejemplo y, cuando lo hemos hecho con oncoSimulPop, la función nueva compositionPop2 ha fallado. Falla a la hora de generar el data frame, que por lo que sea el número de filas no es el mismo que el número de etiquetas (entiendo yo). Revísalo, porque a mí me parece mejor trabajar probablemente con sapply. 

## Ejemplo Álex:

a: beneficio debido a la acidez si hay células que producen ácido.
v: beneficio (para las células aeróbicas) debido a la vascularización si hay células que
  están (sobre) produciendo VEGF a un coste c.
c: coste de producir VEGF.

# Situación 1: 
De acuerdo a la figura 1, en esta primera situación hay oscilaciones dentro de la población, pero debería acabar ganando G.

In this case, we see that if the fitness benefit of a single unit of acidification is higher than the maximum benefit from the club good for A, G will always have a strictly higher fitness than A, and be selected for. In this scenario, the population will converge towards all G, regardless of the initial proportions (as long as there is at least some G in the population).

```{r, message=FALSE}
avc <- function (a, v, c) {
  data.frame(Genotype = c("WT", "G", "V", "A"),
             Fitness = c("1",
                         paste0("1 + ", a, " * (f_1 + 1)"),
                         paste0("1 + ", a, " * f_1 + ", v, " * (f_2 + 1) - ", c),
                         paste0("1 + ", a, " * f_1 + ", v, " * f_2")
                         ))
                          }

afavc <- allFitnessEffects(genotFitness = avc(2.5, 2, 1), # Parametros de la interaccion
                           frequencyDependentFitness = TRUE,
                           frequencyType = "rel")

simulation <- oncoSimulPop(20,
                           mc.cores = 6,
                           afavc,
                           model = "McFL",
                           onlyCancer = FALSE,
                           finalTime = 25,
                           mu = 1e-2,
                           initSize = 4000,
                           keepPhylog = TRUE,
                           seed = NULL,
                           errorHitMaxTries = FALSE,
                           errorHitWallTime = FALSE)

plot(simulation, show = "genotypes", type = "line")
compositionPop2(simulation)

```

# Situación 2: 
De acuerdo a la figura 1, en esta primera situación debería ganar siempre V.

If the benefit to V from their extra unit of vascularisation is higher than the cost c to produce that unit, then V will always have a strictly higher fitness than A. In addition, if the maximum possible benefit of the club good to A is higher than the benefit of an extra unit of acidification, then for sufficiently high number of V, G will have lower fitness than A. When both conditions are satisfied, the population will converge towards all V.

```{r, message=FALSE}
avc <- function (a, v, c) {
  data.frame(Genotype = c("WT", "G", "V", "A"),
             Fitness = c("1",
                         paste0("1 + ", a, " * (f_1 + 1)"),
                         paste0("1 + ", a, " * f_1 + ", v, " * (f_2 + 1) - ", c),
                         paste0("1 + ", a, " * f_1 + ", v, " * f_2")
                         ))
                          }

afavc <- allFitnessEffects(genotFitness = avc(2.5, 7, 1), # Parametros de la interaccion
                           frequencyDependentFitness = TRUE,
                           frequencyType = "rel")

simulation <- oncoSimulPop(20,
                           mc.cores = 6,
                           afavc,
                           model = "McFL",
                           onlyCancer = FALSE,
                           finalTime = 25,
                           mu = 1e-2,
                           initSize = 4000,
                           keepPhylog = TRUE,
                           seed = NULL,
                           errorHitMaxTries = FALSE,
                           errorHitWallTime = FALSE)

plot(simulation[[1]], show = "genotypes", type = "line")
compositionPop2(simulation)

```

# Situación 3: 
De acuerdo a la figura 1, en esta situación debería ganar G.

We know that if the benefit from an extra unit of vascularisation in a fully aerobic group is lower than the cost c to produce that unit, then for a sufficiently low proportion of G and thus sufficiently large number of A sharing the club good, A will have higher fitness than V. This will lead to a decrease in the average fitness of A. A lower fitness in A will lead to an increase in the proportion of G until the aerobic groups (among which the club good is split) get sufficiently small and fitness starts to favour V over A, swinging the dynamics back.

```{r, message=FALSE}
avc <- function (a, v, c) {
  data.frame(Genotype = c("WT", "G", "V", "A"),
             Fitness = c("1",
                         paste0("1 + ", a, " * (f_1 + 1)"),
                         paste0("1 + ", a, " * f_1 + ", v, " * (f_2 + 1) - ", c),
                         paste0("1 + ", a, " * f_1 + ", v, " * f_2")
                         ))
                          }

afavc <- allFitnessEffects(genotFitness = avc(37.5, 2, 1), # Parametros de la interaccion
                           frequencyDependentFitness = TRUE,
                           frequencyType = "rel")

simulation <- oncoSimulPop(20,
                           mc.cores = 6,
                           afavc,
                           model = "McFL",
                           onlyCancer = FALSE,
                           finalTime = 25,
                           mu = 1e-2,
                           initSize = 4000,
                           keepPhylog = TRUE,
                           seed = NULL,
                           errorHitMaxTries = FALSE,
                           errorHitWallTime = FALSE)

plot(simulation, show = "genotypes", type = "line")
compositionPop2(simulation)

```

## Evolutionary games theory: Hawk and Dove example
The example we are going to show is one of the first games that Maynar Smith analyzed. This classic game is: Hawk and Dove (https://en.wikipedia.org/wiki/Chicken_(game))

In this game, the two competitors are subtypes of the same species but with different strategies.  The Hawk first displays aggression, then escalates into a fight until it either wins or is injured (loses). The Dove first displays aggression, but if faced with major escalation runs for safety. If not faced with such escalation, the Dove attempts to share the resource.

![Payoff Matrix for Hawk Dove Game](.\Images_markdown\Hawk_Dove_payoff.PNG)

Given that the resource is given the value V, the damage from losing a fight is given cost C:[1]

- If a Hawk meets a Dove he gets the full resource V to himself
- If a Hawk meets a Hawk – half the time he wins, half the time he loses...so his average outcome is then V/2 minus C/2
- If a Dove meets a Hawk he will back off and get nothing – 0
- If a Dove meets a Dove both share the resource and get V/2

The actual payoff however depends on the probability of meeting a Hawk or Dove, which in turn is a representation of the percentage of Hawks and Doves in the population when a particular contest takes place. That in turn is determined by the results of all of the previous contests. If the cost of losing C is greater than the value of winning V (the normal situation in the natural world) the mathematics ends in an stationary point (ESS), a mix of the two strategies where the population of Hawks is V/C.

In this case we assume a stable equilibrium in the population dynamics, that is, although there are external variations in the model, it recovers and returns to equilibrium (as shown in the image below).

![Stable equilibrium](.\Images_markdown\Stable_equilibrium.jpg)

We are going to simulate with OncoSimulR the situation in which the cost of losing C is greater than the value of gaining V (C = 10, V = 2). Before performing the simulation, let's look at the fitness of each competitor.

```{r, message=F}
## Stablish Genotype-Fitnees mapping. D = Dove, H = Hawk
H_D_fitness <- function(c, v,
                    gt = c("WT", "H", "D")) {
  data.frame(Genotype = gt,
             Fitness = c(
               paste0("1"),
               paste0("1 + f_1 *", (v-c)/2, "+ f_2 *", v),
               paste0("1 + f_2 *", v/2)))
}

## Fitness Effects specification
HD_competition <-allFitnessEffects(
  genotFitness = H_D_fitness(10, 2, 
                         gt = c("WT", "H", "D")),
  frequencyDependentFitness = TRUE,
  frequencyType = "rel",
  spPopSizes = c(5000, 5000, 5000))

## Plot fitness landscape of genotype "H, D" evaluation
data.frame("Doves_fitness" = evalGenotype(genotype = "D", fitnessEffects = HD_competition), 
           "Hawks_fitness" = evalGenotype(genotype = "H", fitnessEffects = HD_competition)) 
```

We observe that the penalty of fighting (C > V) benefits the dove in terms of fitness respect to the hawk. 

``` {r, message=F}
options(stringsAsFactors = FALSE)
## Simulated trajectories
simulation <- oncoSimulPop(2,
                           mc.cores = 6,
                           HD_competition,
                           model = "McFL", # There is no collapse
                           onlyCancer = FALSE,
                           finalTime = 500,
                           mu = 1e-2, # Quick emergence of D and H
                           initSize = 4000,
                           keepPhylog = TRUE,
                           seed = NULL,
                           errorHitMaxTries = FALSE,
                           errorHitWallTime = FALSE)


oldpar <- par (
  # Change the colors
  col.main="black", col.lab="blue",
  # Titles in bold
  font.main=2, font.lab=2,
  # Change font size
  cex.main=2, cex.lab=1.2,
  # Vertically
  las = 1
  )

## Plot first trajectory as an example
plot(simulation[[1]], show = "genotypes", type = "line", xlim = c(40, 500), lwdClone = 2, ylab = "Number of individuals", main = "Hawk and Dove trajectory")

par <- oldpar

## Agreement: Show boxplot from several simulations. 
compositionPop2(simulation)
```

As mentioned above, mathematically when a stationary point (ESS) is reached the relative frequency of hawks is V/C and doves 1-(V/C). Considering f_H as relative frecuency of hawks and f_D = 1-f_H as frequency of doves:
Hawk:
$1 + f_H*(v-c)/2 + (1-f_H)*v$

Dove:
$1 + (1-f_H)*v/2$

Hawk = Dove: 
$1 + f_H*(v-c)/2 + (1-f_H)*v = 1 + (1-f_H)*v/2$ 

Resolving for f_H:
$f_H = v/c$

Therefore, the relative frequency of hawks in equilibrium is equal to V/C. In our case it would be 20% (C = 10, V = 2). Let's check it:

```{r}
## Recover the final result from first simulation
result <- tail(simulation[[1]][[1]], 1)

## Get the number of organisms from each species
n_WT <- result[2]
n_D <- result[3]
n_H <- result[4]
total <- n_WT + n_D + n_H

## Dove percentage
data.frame("Doves" = round(n_D/total, 2)*100, "Hawks" = round(n_H/total, 2)*100  )
```

To sum up, this example shows that when the risks of contest injury or death (the Cost C) is significantly greater than the potential reward (the benefit value V), the stable population will be mixed between aggressors and doves, and the proportion of doves will exceed that of the aggressors. This explains behaviours observed in nature.



### Comentarios
* No sé qué situación es mejor, si mantener WT independiente de S (a pesar de que el programa modela que S aparece como producto mutacional de WT) o no. Utilizando uno u otro los valores de los parámetros cambian para representar un mismo escenario (a pesar de que aquí tienen los mismos, pero para que las gráficas se parecieran más, los valroes deben ser ligeramente distintos). He puesto los dos casos para que comparéis y me digais.
* Creo que cada una de las simulaciones habría que hacerla varias veces, porque los resultados a veces son ligeramente distintos  si lo corres varias veces. 

### Cosas a tener en cuenta
--> Cuando una subpoblación llega a 1e8 se para la simulación

--> Hay que tener en cuenta que si uno de los clones (mutantes) no se origina a partir del WT y el WT muere antes de que pueda aparecer, el programa no lo coge y nos dará error en la simulación repetida 1000 veces. Podemos solucionarlo aumentando la tasa de natalidad del WT o aumentando la tasa de mutación del WT o el tamaño inicial del WT. 

--> spPopSizes no soluciona el problema anterior porque solo lo emplea la función evalAllFitnessEffects. OncoSimulIndv y las otras no lo tienen en cuenta, así que no se soluciona nada. 


--> McFL para cuando no hay colapso (no dependencia total entre clones)
--> McFLD depedencia total (si se muere uno el otro debe acabar muriendo porque no tiene más recursos)

--> Dificultados: cmabiando la leyenda, cambiando los ejes tras recortar la gráfica, plotfitnessLandscape te da combinaciones de mutantes y no me interesa

**Modelo de Lotka-Volterra**
<https://books.google.es/books?id=HgAgMwsYOtkC&pg=PA167&lpg=PA167&dq=a21+a12+Lotka&source=bl&ots=SSTfj46eM6&sig=ACfU3U1aG5qdkQT56tmSrO0SC7SROkKVeQ&hl=es&sa=X&ved=2ahUKEwivovuP-9PmAhWRyoUKHRslAREQ6AEwA3oECAcQAQ#v=onepage&q=a21%20a12%20Lotka&f=false>

### Cosas que mejorar:
1) Graficas: sacar la leyenda y ponerla a mano
2) Intentar poner mas de 2 especies (una movida matematica)